modules/caddyhttp/reverseproxy/reverseproxy.go

```diff
@@ -421,6 +421,12 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request, next caddyht
 			return nil
 		}
 
+		// detect connection timeout to the upstream and respond with 504
+		switch e, ok := proxyErr.(net.Error); ok {
+		case e.Timeout():
+			return caddyhttp.Error(http.StatusGatewayTimeout, proxyErr)
```

Comment:
Returning here bypasses all the health check and load balancing features.

Reply:
Oh, right, because of the retries. Indeed... I'm gonna remove it!